fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/storestate/MetaDataVersionStampStoreStateCache.java

```diff
@@ -0,0 +1,143 @@
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStoreBase;
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.apple.foundationdb.record.provider.foundationdb.SubspaceProvider;
+import com.apple.foundationdb.tuple.ByteArrayUtil;
+import com.google.common.cache.Cache;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An implementation of the {@link FDBRecordStoreStateCache} that uses the database's
+ * {@linkplain FDBRecordContext#getMetaDataVersionStampAsync(IsolationLevel) meta-data version-stamp} key as the
+ * cache invalidation key. This will only cache the meta-data if it is "safe", that is, if the record store
+ * has marked its state as cacheable. This means that one should first set the state cacheability to {@code true}
+ * on any record store whose state one would want cached by calling {@link FDBRecordStore#setStateCacheability(boolean)}.
+ * Then only those stores will be cached by this store state cache.
+ *
+ * @see FDBRecordStore#setStateCacheabilityAsync(boolean)
+ * @see FDBRecordContext#getMetaDataVersionStamp(IsolationLevel)
+ */
+public class MetaDataVersionStampStoreStateCache implements FDBRecordStoreStateCache {
+    @Nonnull
+    private final FDBDatabase database;
+    @Nonnull
+    private final Cache<SubspaceProvider, FDBRecordStoreStateCacheEntry> cache;
+
+    MetaDataVersionStampStoreStateCache(@Nonnull FDBDatabase database, @Nonnull Cache<SubspaceProvider, FDBRecordStoreStateCacheEntry> cache) {
+        this.database = database;
+        this.cache = cache;
+    }
+
+    @Nonnull
+    private FDBRecordStoreStateCacheEntry getNewerEntry(@Nonnull FDBRecordStoreStateCacheEntry entry1, @Nonnull FDBRecordStoreStateCacheEntry entry2) {
+        if (entry1.getMetaDataVersionStamp() == null) {
+            return entry2;
+        } else if (entry2.getMetaDataVersionStamp() == null) {
+            return entry1;
+        } else {
+            return ByteArrayUtil.compareUnsigned(entry1.getMetaDataVersionStamp(), entry2.getMetaDataVersionStamp()) >= 0 ? entry1 : entry2;
+        }
+    }
+
+    private void addToCache(@Nonnull SubspaceProvider subspaceProvider, @Nonnull FDBRecordStoreStateCacheEntry cacheEntry) {
+        cache.asMap().merge(subspaceProvider, cacheEntry, (entry1, entry2) -> {
+            final FDBRecordStoreStateCacheEntry newerEntry = getNewerEntry(entry1, entry2);
+            if (newerEntry.getRecordStoreState().getStoreHeader().getCacheable()) {
+                return newerEntry;
+            } else {
+                return null;
+            }
+        });
+    }
+
+    private void invalidateOlderEntry(@Nonnull SubspaceProvider subspaceProvider, @Nonnull byte[] metaDataVersionStamp) {
+        cache.asMap().computeIfPresent(subspaceProvider, (ignore, existingEntry) -> {
+            // Invalidate the key unless the cached meta-data is newer than this meta-data version stamp
```

Comment:
It's actually newer than or equal to, but I guess equal is impossible with proper usage, isn't it?

Reply:
I updated the comment to match the code.

I think that that's right--that with proper usage, the two would never be equal. In particular, this is only called by someone who thinks that for `metaDataVersionStamp = x`, that the store state isn't cacheable, so if the existing entry had a value equal to `metaDataVersionStamp` and *was* cached, that would mean that the cache has two different ideas as to the state of the meta-data at that meta-data version stamp.

I suppose, in theory, that should throw an error?